spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java

```diff
@@ -0,0 +1,319 @@
+ * The various implementations are registered with the {@link org.apache.iceberg.spark.SparkCatalog}
+ * such that the function can be used as {@code truncate(col, width)} or {@code truncate(col, 2)}.
+ */
+public class TruncateFunction implements UnboundFunction {
+
+  public TruncateFunction() {
+
+  }
+
+  private static void validateSourceTypeUsingSparkTypes(DataType dt) {
+    if (!DataTypes.IntegerType.acceptsType(dt) && !DataTypes.LongType.acceptsType(dt) &&
+        !DataTypes.StringType.acceptsType(dt) && !DataTypes.BinaryType.acceptsType(dt) &&
+        !(dt instanceof DecimalType)) {
+      throw new UnsupportedOperationException(
+          "Cannot truncate input type. Expected one of [IntegerType, LongType, DecimalType, StringType, BinaryType], " +
+              "but found " + dt.catalogString());
+    }
+  }
+
+  private static void validateTruncationWidthType(DataType widthType) {
+    if (!(widthType instanceof IntegerType ||
+        !(widthType instanceof LongType && DataTypes.IntegerType.acceptsType(widthType)))) {
+      throw new UnsupportedOperationException("Expected truncation width to be an integer type");
+    }
+  }
+
+  @Override
+  public BoundFunction bind(StructType inputType) {
+    if (inputType.fields().length != 2) {
+      throw new UnsupportedOperationException(
+          String.format("Invalid input type. Expected 2 fields but found %s", inputType.fields().length));
+    }
+
+    StructField fieldToTruncate = inputType.apply(0);
+    StructField truncationWidth = inputType.apply(1);
+
+    DataType toTruncateSparkType = fieldToTruncate.dataType();
+    Type sourceIcebergType = SparkSchemaUtil.convert(fieldToTruncate.dataType());
+
+    validateSourceTypeUsingSparkTypes(fieldToTruncate.dataType());
+    validateTruncationWidthType(truncationWidth.dataType());
+
+    switch (sourceIcebergType.typeId()) {
+      case INTEGER:
+        return new TruncateInteger();
+      case LONG:
+        return new TruncateLong();
+      case DECIMAL:
+        return new TruncateDecimal(
+            ((DecimalType) toTruncateSparkType).precision(),
+            ((DecimalType) toTruncateSparkType).scale());
+      case STRING:
+        return new TruncateString();
+      case BINARY:
+        return new TruncateBinary();
+      default:
+        throw new UnsupportedOperationException(
+            "Cannot truncate type: " + sourceIcebergType.typeId());
+    }
+  }
+
+  @Override
+  public String description() {
+    return "Truncate - The Iceberg truncate function used for truncate partition transformations.";
+  }
+
+  @Override
+  public String name() {
+    return "truncate";
+  }
+
+  public abstract static class TruncateBase<T> implements ScalarFunction<T> {
+    @Override
+    public String name() {
+      return "truncate";
+    }
+
+    @Override
+    public boolean isResultNullable() {
+      return true;
+    }
+
+    @Override
+    public boolean isDeterministic() {
+      return true;
```

Comment:
`true` is the default for both `isDeterministic` and `isResultNullable`, so these implementations shouldn't be necessary.

Reply:
Removed the unnecessary overrides.